### ISSUE TITLE
make pmap compilation cache "act globally" with no axis name

### ISSUE THREAD

**Issue by mattjj** [google/jax#1600]
Both the `jit` and `pmap` compilation caches are global. The `jit` cache acts that way:

```python
from jax import jit

def f(x):
  print("miss")
  return x

jit(f)(3)  # prints
jit(f)(3)  # doesn't print, i.e. cache hit
```

But the `pmap` one seems not to when you don't provide a name for the axis:

```python
from jax import pmap
from jax.lib import xla_bridge as xb
import jax.numpy as np

def f(x):
  print("miss")
  return x

x = np.ones(xb.device_count())
pmap(f)(x)  # prints
pmap(f)(x)  # prints

pmap(f, axis_name='i')(x)  # prints
pmap(f, axis_name='i')(x)  # doesn't print
```

The issue is that when an axis name isn't provided, we create a dummy one, and we create a new one each time we evaluate `pmap(f)` to produce a new pmapped version of `f`. We should probably remedy that, since patterns like

```python
for i in range(100):
  x = pmap(update)(x)
```

are handy, but they're currently a foot-gun with respect to compilation caching.
